### PR TITLE
Slim down container image

### LINKED ISSUE #9: Slim down container image
Currently the size of the geoCML container is floating at around ~~4GB~~ 6GB. We could reduce the size of the container by removing some un-needed dependencies from the image; we install packages that are not needed for doing GIS work. Some dependencies that can be removed include:

- VS Code
- Google Chrome
- Git

We should also take some time to review our Ansible playbook and determine where we can refactor playbook.yml to decrease build times.

### PR DESCRIPTION
**Pull request by TristanDamron**
Resolves #9 

This PR aims to slim down the size of geoCML containers by removing dependencies that are not necessary for doing GIS work. In addition to reducing the size of our container image, this PR will decrease container build times for users building locally and  our CI/CD.

### CODE REVIEW

**Comment by TristanDamron** [comment]
@capsulecorplab This PR is ready for you to review! 

There's still some more that I want to do as far as making the container more lightweight, but I think we should handle that with another PR. Some ideas I want to explore in the future include:
- multi-stage builds
- creating a custom, slimmed down Kasm image

File: Dockerfile [hunk 1]
```diff
@@ -10,15 +10,32 @@ WORKDIR $HOME
 
 # copy over install_files/ for use in playbooks
 ADD install_files $HOME/install_files
+RUN apt update && apt install -y sudo
+
+# fix for .gnupg/ permissions when building custom images
+RUN apt install -y qgis
```
Comment (capsulecorplab) [inline Dockerfile:16]: Any reason we're installing qgis (as well as postgresql, postgis, pg admin, & python packages) in the Dockerfile as opposed to in the ansible playbook?

File: requirements.yaml [hunk 2]
```diff
@@ -1,11 +1,5 @@
 ---
 # Add Ansible Galaxy Packages here, role_example_hello included to prevent errors in template testing
-roles:
-- irixjp.role_example_hello           # https://galaxy.ansible.com/irixjp/role_example_hello
```
Comment (capsulecorplab) [inline requirements.yaml:4]: We should keep the hello world example, as it's a safe measure for debugging ansible

File: playbook.yaml [hunk 3]
```diff
@@ -21,104 +21,38 @@
...
-    lineinfile:
-      path: "/etc/skel/.profile"
-      line: 'export PS1="\[\033[00;32m\][\w]\$\[\033[00m\] "'
-
-  # Handlers: the list of handlers that are executed as a notify 
-  #           key from a task
-
-  # Roles: list of roles to be imported into the play
-  roles:
-  - role: irixjp.role_example_hello
```
Comment (capsulecorplab) [inline playbook.yaml:48]: We should keep this as a debugging measure for ansible

File: Dockerfile (hunk 1 above)
Comment (TristanDamron) [reply Dockerfile:16]: Ansible was a massive bottleneck for us the way we were previously using it. The way we're using it now, Dockerfiles allow us to cache each installation step in a layer whereas previously the entire Ansible playbook ran in a single layer. This meant that for any small change we made to the playbook, Docker would have to rebuild that entire layer. By separating this out we're increasing the number of layers but we're saving a lot of time as far as building the container is concerned because each of those layers can be cached. IMO we should never use Ansible for installation, and only rely on it for more complex configurations that would be too difficult to do with a Dockerfile. 

tl;dr Dockerfiles are much better as far as installing software packages versus Ansible for our needs

File: requirements.yaml (hunk 2 above)
Comment (TristanDamron) [reply requirements.yaml:4]: What value does this role provide users building the container that they wouldn't already get by running the playbook and checking the results?

Comment (capsulecorplab) [reply requirements.yaml:4]: If there's an issue building with the playbook, it'd be good to know whether it's an ansible issue or an issue specific to the thing we're installing. The hello example would at least tell us ansible seems to be working fine (or not)

Comment (TristanDamron) [reply requirements.yaml:4]: The traceback would tell us why the failure happened, right? I think it would be pretty easy to tell if it's failing from an internal Ansible issue versus a bad play in the playbook. That being said, I can understand your perspective on wanting to include this role. I guess I'm not completely opposed to putting it back in the image, but I want to make sure we _absolutely need it_ before I do.

File: Dockerfile (hunk 1 above)
Comment (capsulecorplab) [reply Dockerfile:16]: Fair points; let's use Dockerfile for installations & Ansible playbooks for managing configuration

File: requirements.yaml (hunk 2 above)
Comment (TristanDamron) [reply requirements.yaml:4]: Per our conversation earlier, fe7257d readds the hello world role but on the condition that it's commented out until it's needed.

File: requirements.yaml [hunk 4]
```diff
@@ -2,10 +2,6 @@
 # Add Ansible Galaxy Packages here, role_example_hello included to prevent errors in template testing
 roles:
 - irixjp.role_example_hello           # https://galaxy.ansible.com/irixjp/role_example_hello
```
Comment (capsulecorplab) [inline requirements.yaml:4]: Should comment this out too

Comment (TristanDamron) [reply requirements.yaml]: Resolved with ea41eb1

**Comment by TristanDamron** [comment]
@capsulecorplab I've got some additional metrics related to this PR that you might find interesting :smile: 

Previous CI builds off of main took around 18 minutes, with this update our CI runs in under 9. We've effectively cut our build times in half!

**Comment by capsulecorplab** [comment]
> @capsulecorplab I've got some additional metrics related to this PR that you might find interesting smile
> 
> Previous CI builds off of main took around 18 minutes, with this update our CI runs in under 9. Without caching anything, we've effectively cut our build times in half!

Nice!

**Comment by capsulecorplab** [comment]
Just pulled the latest image https://github.com/users/capsulecorplab/packages/container/kasm-gis-workspace/35132314?tag=sha-ea41eb1 built from my forked repo https://github.com/capsulecorplab/kasm-gis-workspace/runs/7866121766?check_suite_focus=true. Anything I should test before approving the PR?

**Comment by TristanDamron** [comment]
If you want to do some basic regression, go ahead! There should be no functional changes to Postgres and QGIS.

**Comment by TristanDamron** [comment]
@capsulecorplab Is this ready to be merged?

**Comment by capsulecorplab** [comment]
Sorry for the delay! Been out traveling. Where can I find documentation on running a regression example?

**Comment by TristanDamron** [comment]
This site is probably a good place to start: https://www.qgistutorials.com/en/

**Comment by capsulecorplab** [comment]
[Map making](https://www.qgistutorials.com/en/docs/3/making_a_map.html) seems to be working! Though I messed up somewhere between step 14-15 :sweat_smile:  Making_a_map_-_Japan

**Review by capsulecorplab — APPROVED** [review]
LGTM!
